### PR TITLE
Update lints to embark lints 0.3

### PR DESCRIPTION
**Pull request by DeanBDean**
Fixes https://github.com/EmbarkStudios/tame-gcs/issues/45

One of the biggest changes impacts tests that have been ignored. I tried running them by removing the ignore, but they seem to rely on an ENV VAR that I don't have (TAME_GCS_TEST_SVC_ACCOUNT, TAME_GCS_TEST_BUCKET, TAME_GCS_TEST_OBJECT). These seem like things that maybe run in CI, but I couldn't see any references to that in the repo.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -36,6 +36,7 @@ base64 = "0.13" # Follows version used by rustls
 bytes = "1.0" # Follows version used by http
 chrono = { version = "0.4", features = ["serde"] }
 futures-util = { version = "0.3", optional = true, features = ["io"] }
+hex = "0.4"
```
Comment (Jake-Shadle) [inline Cargo.toml:39]:
```suggestion
```

There is no need to add a dependency and remove the util code in a lint related PR.

File: src/error.rs [hunk 2]
```diff
@@ -45,6 +45,14 @@ pub enum Error {
...
 }
 
+// impl std::error::Error for Error {
+//     fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+//         match self {
+//             Error::InvalidCharacterCount(, )
+//         }
+//     }
+// }
+
```
Comment (Jake-Shadle) [inline src/error.rs:55]:
```suggestion
```

No need to keep this if it's not being used.

File: src/response.rs [hunk 3]
```diff
@@ -73,7 +73,7 @@ where
         }
     }
 
-    /// Try to get an http::Response
+    /// Try to get an `http::Response`
```
Comment (Jake-Shadle) [inline src/response.rs:76]:
```suggestion
    /// Try to get an [`http::Response`]
```

File: src/signed_url.rs [hunk 4]
```diff
@@ -18,7 +18,7 @@ pub struct UrlSigner<D, S> {
 
 #[cfg(feature = "signing")]
 impl UrlSigner<signing::RingDigest, signing::RingSigner> {
-    /// Creates a UrlSigner implemented via `ring`
+    /// Creates a `UrlSigner` implemented via `ring`
```
Comment (Jake-Shadle) [inline src/signed_url.rs:21]:
```suggestion
    /// Creates a [`UrlSigner`] implemented via `ring`
```

File: src/signed_url.rs [hunk 5]
```diff
@@ -29,7 +29,7 @@ where
     D: signing::DigestCalulator,
     S: signing::Signer,
 {
-    /// Creates a new UrlSigner from a `DigestCalculator` implementation
+    /// Creates a new `UrlSigner` from a `DigestCalculator` implementation
```
Comment (Jake-Shadle) [inline src/signed_url.rs:32]:
```suggestion
    /// Creates a new [`UrlSigner`] from a [`DigestCalculator`] implementation
```

File: src/signed_url.rs [hunk 6]
```diff
@@ -112,6 +112,10 @@ where
                     }
 
                     key_vals.push_str(
+                        #[allow(clippy::map_err_ignore)]
+                        // Allowing because the intent is to ignore the `ToStrError`, and because the
+                        // new error requires the HeaderValue from this context, and it is uglier to
+                        // create a conversion function that accounts for that than to ignore the lint
                         val.to_str()
                             .map_err(|_| Error::OpaqueHeaderValue(val.clone()))?,
```
Comment (Jake-Shadle) [inline src/signed_url.rs:120]:
```suggestion
                        val.to_str()
                            .map_err(|_err| Error::OpaqueHeaderValue(val.clone()))?,
```

File: src/types.rs [hunk 7]
```diff
@@ -11,7 +11,7 @@ pub struct BucketName<'a> {
 }
 
 impl<'a> BucketName<'a> {
-    /// Creates a BucketName without validating it, meaning
+    /// Creates a `BucketName` without validating it, meaning
```
Comment (Jake-Shadle) [inline src/types.rs:14]:
```suggestion
    /// Creates a [`BucketName`] without validating it, meaning
```

File: src/types.rs [hunk 8]
```diff
@@ -149,19 +148,24 @@ impl<'a> ObjectName<'a> {
...
+                | '['
+                | ']'
+                | '*'
+                | '?'
+                | '\u{7F}'..='\u{84}'
+                | '\u{86}'..='\u{9F}' => {}
                 // Avoid using control characters that are illegal in XML 1.0 (#x7F–#x84 and #x86–#x9F):
                 // these characters will cause XML listing issues when you try to list your objects.
-                '\u{7F}'..='\u{84}' => {}
-                '\u{86}'..='\u{9F}' => {}
```
Comment (Jake-Shadle) [inline src/types.rs:164]: I'd like the warning to be ignored here, each section has its own preceding comment about why they are not allowed, this loses that.

File: src/util.rs [hunk 9]
```diff
@@ -1,28 +1,9 @@
...
-
-    unsafe { Some(str::from_utf8_unchecked(&output[0..input.len() * 2])) }
-}
-
 pub fn get_content_length(headers: &http::HeaderMap) -> Option<usize> {
     headers.get(http::header::CONTENT_LENGTH).and_then(|h| {
+        #[allow(clippy::map_err_ignore)]
+        // Allowing because the intent is to swallow the error
         h.to_str()
             .map_err(|_| ())
```
Comment (Jake-Shadle) [inline src/util.rs:8]:
```suggestion
        h.to_str()
            .map_err(|_err| ())
```

**Review by Jake-Shadle — CHANGES_REQUESTED** [review]
Thanks for the PR! The biggest thing is that I don't see any reason for adding a new dependency as it's completely unrelated to the lints this PR is adding, so please reverf that.

**Comment by DeanBDean** [comment]
Unfortunately, if I just revert that the lint's won't pass. It was replacing unsafe code with (I am pretty sure, I can't double check on my phone) safe code. It didn't seem like the kind of code that necessitated ignoring the lint, but also reimplementing the solution in that crate also seemed like overkill. If that unsafe code is needed I can always ignore the lint

**Comment by Jake-Shadle** [comment]
Yes, it's fine to turn off the lint in just that one function.

File: src/error.rs (hunk 2 above)
Comment (DeanBDean) [reply src/error.rs]: Sorry, this was an oversight on my part

**Comment by DeanBDean** [comment]
@Jake-Shadle So, I couldn't quite bring myself to allowing the no unsafe lint without one more shot. This time I mostly kept the original code while modifying it based on that other package. Let me know if that works

**Review by Jake-Shadle — CHANGES_REQUESTED** [review]
Nice thanks!

**Review by Jake-Shadle — APPROVED** [review]

**Pull request merged**